### PR TITLE
Lection1

### PR DESCRIPTION
**Pull request by VictorLVA**

### CODE REVIEW

File: Lecture2/Task1/Task1L2.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import java.util.Scanner;
+
+public class Task1L2 {
+    public static void main ( String[] args ) {
+        Scanner in = new Scanner ( System.in );
+        System.out.print ( "Please enter a sentence: " );
+        String sentence = in.nextLine ();
+        String reverse = new StringBuffer(sentence).reverse().toString();
```
Comment (dihnatsyeu) [inline Lecture2/Task1/Task1L2.java]: good choice :) However i would like you to do it in another way, without using java internal mechanism.

File: Lecture2/Task1/Task1L2.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        System.out.println ("Lets reverse: \""+reverse+"\"");
+        String[] words = sentence.split(" ");
+        System.out.println ("Lets split by words:");
+        for(String eachWord : words)
+            System.out.println("\""+eachWord+"\"");
+        System.out.println ("Lets replace spaces with *: \""+sentence.replace ( ' ', '*' )+"\"");
+        System.out.println ("Lets rewrite it with uppercase: \""+sentence.toUpperCase ()+"\"");
+        if (sentence.length () < 10)
+            System.out.println ("Lets get the 5-10 symbols: \""+sentence.substring ( 4 )+"\"");
+        else System.out.println ("Lets get the 5-10 symbols: \""+sentence.substring ( 4, 10 )+"\"");
```
Comment (dihnatsyeu) [inline Lecture2/Task1/Task1L2.java]: when you do substring, last index is exlcuded. you need to have 11 as last index

Comment (dihnatsyeu) [reply Lecture2/Task1/Task1L2.java]: even if you have only 1 condition for if block, surround it with {} brackets to make your code more readable

File: Lecture2/Task2/Task2L2.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        System.out.print ( "Please enter an index of a day within a week (1-7): " );
+        byte day;
+        try {
+            day = in.nextByte ( );
+        }
+        catch (Exception e) {
+            System.out.println ( MESSAGE_INVALID );
+            return;
+        }
+        if (day == 1)
```
Comment (dihnatsyeu) [inline Lecture2/Task2/Task2L2.java]: think how you can achieve the task using array

File: Lecture2/Task2/Task2L2.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
+import java.util.Scanner;
+
+public class Task2L2 {
+
+    private static final String MESSAGE_SELECTED = "You have selected a ";
+    private static final String MESSAGE_INVALID = "You have entered an invalid day's index";
+
+    public static void main ( String[] args ) {
+        Scanner in = new Scanner ( System.in );
```
Comment (dihnatsyeu) [inline Lecture2/Task2/Task2L2.java]: fix formatting. You don't need spaces before and in brackets, i.e. should be:
Scanner in = new Scanner(System.in);
The same is for day = in.nextByte ( ); and below

File: Lecture2/Task5/Task5L2.java [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        boolean checkResultName = false;
+        boolean checkResultLastname = false;
+        while (!checkResultName || !checkResultLastname) {
+            System.out.print ( "\nYour choice: " );
+            String choice = reader.readLine ( );
+            if (listOfWords.contains ( choice ) || listOfLetters.contains ( choice )) {
+                System.out.println ( MESSAGE_AGAIN );
+                continue;
+            }
+            if ((Arrays.equals ( nameArray , starNameArray ) && choice.equalsIgnoreCase ( name )) || (Arrays.equals ( lastnameArray , starLastnameArray ) && choice.equalsIgnoreCase ( lastname ))) {
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: extract this long condition to readable method with boolean return type.

File: Lecture2/Task5/Task5L2.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                System.out.println ( MESSAGE_AGAIN );
+                continue;
+            }
+            if (choice.length ( ) != name.length ( ) && choice.length ( ) != lastname.length ( ) && choice.length ( ) != 1 || !choice.matches ( REGEX_NAME_LASTNAME )) {
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: extract this long condition to readable method with boolean return type.

File: Lecture2/Task5/Task5L2.java [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+                System.out.println ( "You have to enter one letter or a whole name/lastname. Please try again." );
+                continue;
+            }
+            if (choice.length ( ) != 1)
+                wordProcessing ( nameArray , starNameArray , choice , name , lastnameArray , starLastnameArray , lastname , listOfWords );
+            if (choice.length ( ) == 1)
+                letterProcessing ( nameArray , starNameArray , choice , lastnameArray , starLastnameArray , listOfLetters );
+            checkResultName = Arrays.equals ( nameArray , starNameArray );
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: unnecessary validation. You can assign checkResultName/checkResultLastName in the appropriate wordMoveValidation  method.

File: Lecture2/Task5/Task5L2.java [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+    private static final String REGEX_NAME_LASTNAME = "[a-zA-Z]+";
+    private static final String MESSAGE_AGAIN = "Again? Really? Try something else :)";
+
+    private static BufferedReader reader = new BufferedReader ( new InputStreamReader ( System.in ) );
+
+    public static
+    void main ( String[] args ) throws Exception {
+        System.out.println ( "Let's start the game!!!" );
+        Console console = System.console ( );
+        String name = null;
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java:14]: Please, read task carefully. You need to guess your own name. No need to get firstName and lastName from the user's input

File: Lecture2/Task5/Task5L2.java [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+        List <String> listOfLetters = new ArrayList <> ( );
+        char[] nameArray = name.toCharArray ( );
+        char[] lastnameArray = lastname.toCharArray ( );
+        char[] starNameArray = new char[ name.length ( ) ];
+        for ( int i = 0 ; i < starNameArray.length ; i++ )
+            starNameArray[ i ] = '*';
+        char[] starLastnameArray = new char[ lastname.length ( ) ];
+        for ( int i = 0 ; i < starLastnameArray.length ; i++ )
+            starLastnameArray[ i ] = '*';
+        boolean checkResultName = false;
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: boolean names could be more accurate, like isFNameOpened, isLNameOpened

File: Lecture2/Task5/Task5L2.java [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+    }
+
+    private static boolean wordMoveValidation ( char[] array , char[] starArray ) {
+        System.arraycopy ( array , 0 , starArray , 0 , array.length );
+        return true;
+    }
+
+    private static void wordProcessing ( char[] nameArray , char[] starNameArray , String choice , String name , char[] lastnameArray , char[] starLastnameArray , String lastname , List <String> listOfWords ) {
+        boolean checkRightMove = false;
+        if (!Arrays.equals ( nameArray , starNameArray ) && choice.length ( ) == name.length ( ) && choice.equalsIgnoreCase ( name ))
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: Do you really need this validation in that method:!Arrays.equals ( nameArray , starNameArray )? As I see in your code, this method is called only if words are not opened.

File: Lecture2/Task5/Task5L2.java [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+            checkRightMove = wordMoveValidation ( nameArray , starNameArray );
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: not good name for variable, checkRightMove. isWordOpened -> more or less reasonable.

File: Lecture2/Task5/Task5L2.java [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+        letterMessageValidation ( checkRightMoveName , checkRightMoveLastname , nameArray , lastnameArray , starNameArray , starLastnameArray );
+    }
+
+    private static void wordMessageValidation ( boolean checkRightMove , char[] starNameArray , char[] starLastnameArray ) {
+        if (checkRightMove)
+            printWholeWord ( starNameArray , starLastnameArray );
+        else printWrongMove ( );
+    }
+
+    private static boolean wordMoveValidation ( char[] array , char[] starArray ) {
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: no real sense in this method.  If this method did actual array copying, then yes - method is reasonable

File: Lecture2/Task5/Task5L2.java [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    private static final String MESSAGE_AGAIN = "Again? Really? Try something else :)";
+
+    private static BufferedReader reader = new BufferedReader ( new InputStreamReader ( System.in ) );
+
+    public static
+    void main ( String[] args ) throws Exception {
+        System.out.println ( "Let's start the game!!!" );
+        Console console = System.console ( );
+        String name = null;
+        while (name == null)
```
Comment (dihnatsyeu) [inline Lecture2/Task5/Task5L2.java]: you can reduce number of parameters that are required to be sent to the methods by assigning it to the class.

File: Lecture2/Task1/Task1L2.java (hunk 1 above)
Comment (VictorLVA) [reply Lecture2/Task1/Task1L2.java]: fixed

File: Lecture2/Task1/Task1L2.java (hunk 2 above)
Comment (VictorLVA) [reply Lecture2/Task1/Task1L2.java]: - when you do substring, last index is exlcuded. you need to have 11 as last index - FAD
- even if you have only 1 condition for if block, surround it with {} brackets to make your code more readable - fixed

File: Lecture2/Task2/Task2L2.java (hunk 3 above)
Comment (VictorLVA) [reply Lecture2/Task2/Task2L2.java]: fixed

File: Lecture2/Task2/Task2L2.java (hunk 4 above)
Comment (VictorLVA) [reply Lecture2/Task2/Task2L2.java]: fixed

File: Lecture2/Task5/Task5L2.java (hunk 8 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java:14]: just for improvement, not interesting to play with the same words all time :)

File: Lecture2/Task5/Task5L2.java (hunk 9 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: fixed

File: Lecture2/Task5/Task5L2.java (hunk 11 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: fixed

File: Lecture2/Task5/Task5L2.java (hunk 10 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: not need it
fixed

Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: not needed
fixed

File: Lecture2/Task5/Task5L2.java (hunk 12 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: method name fixed

File: Lecture2/Task5/Task5L2.java (hunk 7 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: FAD - this validation for the game ending
wordMoveValidation - only for word processing

but game can be ended by letter

File: Lecture2/Task5/Task5L2.java (hunk 5 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: fixed

File: Lecture2/Task5/Task5L2.java (hunk 6 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: fixed

File: Lecture2/Task5/Task5L2.java (hunk 13 above)
Comment (VictorLVA) [reply Lecture2/Task5/Task5L2.java]: fixed

**Pull request merged**